src/Symfony/Component/Serializer/Serializer.php

```diff
@@ -51,46 +51,42 @@ class Serializer implements SerializerInterface, NormalizerInterface, Denormaliz
      */
     public const EMPTY_ARRAY_AS_OBJECT = 'empty_array_as_object';
 
-    private const SCALAR_TYPES = [
-        'int' => true,
-        'bool' => true,
-        'float' => true,
-        'string' => true,
-    ];
+    private NormalizerInterface $normalizer;
+
+    private DenormalizerInterface $denormalizer;
 
     protected ChainEncoder $encoder;
 
     protected ChainDecoder $decoder;
 
-    /**
-     * @var array<string, array<string, array<bool>>>
-     */
-    private array $denormalizerCache = [];
-
-    /**
-     * @var array<string, array<string, array<bool>>>
-     */
-    private array $normalizerCache = [];
-
     /**
      * @param array<NormalizerInterface|DenormalizerInterface> $normalizers
      * @param array<EncoderInterface|DecoderInterface>         $encoders
      */
     public function __construct(
-        private array $normalizers = [],
+        array $normalizers = [],
         array $encoders = [],
+        ChainNormalizerInterface $normalizerAggregator = null,
```

Comment:
Does it make sense to specify `NormalizerInterface` instead?

IMHO, at some point, the very best signature of that constructor should be:
```php
public function __construct(
  NormalizerInterface $normalizer,
  DenormalizerInterface $denormalizer,
  EncoderInterface $encoder,
  DecodeInterface $decoder,
) {}
```

Indeed, I do think that the fact that normalizers or encoders are "chaining" should not be known by the serializer.

By doing that, we could get rid of the `ChainNormalizerInterface` and `ChainDenormalizerInterface`.
And we can harmonize the normalization and the encoding.

Reply:
I tend to agree that the `chain` aspect should remain an implementation detail

Reply:
Okey. Cool. I'll try to see if it make sense.

Reply:
It did!